sql/analyzer/optimization_rules.go

```diff
@@ -236,24 +148,37 @@ func nodeSources(node sql.Node) []string {
 	return result
 }
 
-// expressionSources returns the set of sources from any GetField expressions in the expression given.
-func expressionSources(expr sql.Expression) []string {
+// expressionSources returns the set of sources from any GetField expressions
+// in the expression given, and a boolean indicating whether the expression
+// is null rejecting from those sources.
+func expressionSources(expr sql.Expression) ([]string, bool) {
```

Comment:
allowNulls is probably more natural here

Beware bools with negations in the name

Reply:
nullRejecting is the standardized terminology in relational algebra afaik